SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+          * mask
+          * extra_coords
+
+        the following attributes must be copied unaltered to the returned NDCube:
+
+          * unit
+          * meta
+
+        """
+
+    @abc.abstractmethod
+    @property
+    def extra_coords(self) -> ExtraCoordsABC:
+        """
+        Coordinates not decribed by ``NDCubeABC.wcs`` which vary along one or more axes.
+        """
+
+    @abc.abstractmethod
+    @property
+    def global_coords(self) -> GlobalCoordsABC:
+        """
+        Coordinate metadata which applies to the whole cube.
+        """
+
+    @abc.abstractmethod
+    @property
+    def wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube.
+
+        If a `BaseLowLevelWCS` is provided to the constructor it should be
+        "upgraded" to a high level object.
+        """
+
+    @abc.abstractmethod
+    @property
+    def combined_wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube, including the coordinates specified in ``.extra_coords``.
+
+        This transform should implement the high level wcsapi, and have
+        `pixel_n_dim` equal to the number of array dimensions in the
+        `.NDCube`. The number of world dimensions should be equal to the
+        number of world dimensions in ``self.wcs`` and in ``self.extra_coords`` combined.
+        """
+
+    @abc.abstractmethod
+    @property
+    def array_axis_physical_types(self) -> Iterable[Tuple[str, ...]]:
+        """
+        Returns the WCS physical types associated with each array axis.
+
+        Returns an iterable of tuples where each tuple corresponds to an array axis and
+        holds strings denoting the WCS physical types associated with that array axis.
+        Since multiple physical types can be associated with one array axis, tuples can
+        be of different lengths. Likewise, as a single physical type can correspond to
+        multiple array axes, the same physical type string can appear in multiple tuples.
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords(self,
+                          *axes: Union[int, str],
+                          edges: bool = False,
+                          wcs: Optional[Union[BaseHighLevelWCS, ExtraCoordsABC]] = None
+                          ) -> Iterable[Any]:
+        """
+        Returns world coordinate values of all pixels for all axes.
+
+        Parameters
+        ----------
+        axes
+            Axis number in numpy ordering or unique substring of
+            `.NDCube.wcs.world_axis_physical_types` or
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
+            The edges argument helps in returning ``pixel_edges``
+            instead of ``pixel_values``. Default value is False,
+            which returns ``pixel_values``. True return ``pixel_edges``
+
+        wcs
```

Comment:
Why allow passing an arbitrary `wcs`, when this object has `self.wcs`?

Reply:
This isn't really intended to be a truly arbitary WCS (although in theory it could be), it's really intended to be one of `self.wcs`, `self.combined_wcs` or `self.extra_coords`, as a way to limit which coords you are interested in.

Reply:
What's the motivation for allowing that, instead of just doing it on `self.combined_wcs`, and letting the user use the coordinates they want from the full output?

Reply:
This wcs flag is a bit of an afterthought, mainly added due to implementation concerns with `combined_wcs`. It makes less sense for `axis_world_coords[_values]` than it does for `crop`, because there is a simple way to restrict the return anyway. The only real motivation for it being here is it currently has a pretty large performance overhead.

[for crop using `combined_wcs` is very clunky as there is currently no de-duplication of world coordinates when running through `wcs.world_to_pixel`, so you have to specify *all* the world coordinates even if that's more than the number of pixel dimensions.]

In the discussions @DanRyanIrish and I have had about the `wcs=` argument, we have posited it as something which could be considered for deprecation when the implementation of `combined_wcs` is better.

Reply:
I'm still confused; can you give a concrete example of when and how one would use this?

Reply:
I have added a paragraph discussing this further up, I suggest further discussions on the merit of `combined_wcs` go there.